pkg/library/projects/util.go

```diff
@@ -0,0 +1,37 @@
+//
+// Copyright (c) 2019-2023 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package projects
+
+import (
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+)
+
+// Returns all projects (including starter projects) present in a DevWorkspaceTemplateSpec
+func GetAllProjects(workspace *dw.DevWorkspaceTemplateSpec) []dw.Project {
+	projects := workspace.Projects
+
+	// Note: starter project does not allow for specifying clonePath
+	for _, starterProject := range workspace.StarterProjects {
+
+		project := dw.Project{
+			Name:          starterProject.Name,
+			Attributes:    starterProject.Attributes,
+			ProjectSource: starterProject.ProjectSource,
+		}
+		projects = append(projects, project)
+	}
+	return projects
+}
```

Comment:
I'm not sure this is correct behavior, in-context (haven't had a chance to verify though). 

If a resolved workspace ends up with e.g. no projects and two starterProjects, with the second one being the "selected" project, how does the caller know that e.g. the first project will not be cloned but the second one will?

Reply:
You're correct - there's a conflict with the "selected" starter project and how [`getProjectSourcePath()`](https://github.com/devfile/devworkspace-operator/blob/main/pkg/library/container/mountSources.go#L90-L101) only selects the first available project to set $PROJECT_SOURCE. 

I wasn't sure if this rule of "the first available project is what $PROJECT_SOURCE maps to" was documented somewhere, and I couldn't find it, or if it was just a decision that made sense at the time. 

I'll rework things to respect the selected starter project attribute.

Reply:
The env var is (poorly) documented as part of the Devfile schema, in `.commands.exec.commandline`:

> $PROJECT_SOURCE: A path to a project source ($PROJECTS_ROOT/<project-name>). If there are multiple projects, this will point to the directory of the first one.

Reply:
Thank you for clarifying on the documentation of `$PROJECT_SOURCE` @amisevsk :pray: 

@l0rd do you have any input on this? For context:

Reply:
If the attribute `controller.devfile.io/use-starter-project` is set, then we should configure PROJECT_SOURCE accordingly, otherwise we should fall back on the first project.

[And Eclipse Che dashboard is currently setting the attribute at startup](https://github.com/eclipse/che/issues/22486) (without interaction).